test/transaction_test.dart

```diff
@@ -633,5 +633,16 @@ void main() {
       });
       expect(result, []);
     });
+
+    test('can start transactions manually', () async {
+      await conn.execute('BEGIN');
+      await conn.execute(
+        'INSERT INTO t VALUES (@a)',
+        substitutionValues: {'a': 123},
+      );
+      await conn.execute('ROLLBACK');
+
+      await expectLater(conn.query('SELECT * FROM t'), completion(isEmpty));
+    });
```

Comment:
@simolus3 and @isoos is it possible to run queries inside the transaction? Something like this:

```dart
test('some tests in a transaction', () {
  await conn.execute('begin');
  await conn.execute('insert somethings');
  var results = await conn.query('select somethings');
  expect(results.length, 2);
  await conn.execute('rollback');
});
```

The reason why is because I'm running each tests inside transactions to increase test performance significantly. Laravel framework does this in their tests and it works perfectly. I've being using this approach for years and it is a very good feature during testing. This is a real example of how I can run the tests on Tunder framework (for dart):

```dart
main() {
  group('Some tests', () {
    useDatabaseTransactions();
    
    test('this test will execute inside a transction so you can follow the AAA approach', () async {
      // Arrange
      ContactFactory().createMany(10);
      
      // Act
      var contacts = Query<Contact>().paginate(5);
      
      // Assert
      expect(contacts.data.length, 5);
      expect(contacts.meta.perPage, 5);
    });
  });
}
```

We have many benefits of this approach:
1. Faster queries because they're executed in memory and not really persisted in the hard drive or SSD. So no heavy IO operations.
2. Cleaner tests because we don't need to cleanup the database after each test to make them isolated
3. Each test runs completely isolated from contexts because after each test the transaction is rolled back.

If this is not possible right now, how hard would be to implement something like this?

I'm really really looking forward to work with this. If this is possible I'll be definitely migrating it to this package. And thank you so much for all of you guys maintaining this package.

Reply:
In other projects we are using `testWithX` methods, e.g. `testWithDbTransactions`, where the helper method sets up the test, prepends it with DB connection initialization and cleanup afterwards. It is nice as the test's inner method is a callback, so you can do anything before/after easily.

For this specific case, I'm not familiar what the contactfactory does, but it looks like as a global variable, which will usually cause conflicts with tests that may be running concurrently.

Reply:
Great! Thank you for sharing that. It seems it's a pattern in Dart community. I'm reading the tests to understand the behaviour of the package. Thank you for your help @isoos 👍🏻